storagenode/piecestore/endpoint.go

```diff
@@ -322,44 +316,41 @@ func (endpoint *Endpoint) Upload(stream pb.Piecestore_UploadServer) (err error)
 		}
 
 		if message.Done != nil {
-			expectedHash := pieceWriter.Hash()
-			if err := endpoint.VerifyPieceHash(ctx, limit, message.Done, expectedHash); err != nil {
+			calculatedHash := pieceWriter.Hash()
+			if err := endpoint.VerifyPieceHash(ctx, limit, message.Done, calculatedHash); err != nil {
 				return err // TODO: report grpc status internal server error
 			}
 			if message.Done.PieceSize != pieceWriter.Size() {
 				return ErrProtocol.New("Size of finished piece does not match size declared by uplink! %d != %d",
-					message.Done.GetPieceSize(), pieceWriter.Size())
-			}
-
-			if err := pieceWriter.Commit(ctx); err != nil {
-				return ErrInternal.Wrap(err) // TODO: report grpc status internal server error
+					message.Done.PieceSize, pieceWriter.Size())
 			}
 
-			// TODO: do this in a goroutine
 			{
-				// TODO: maybe this should be as a pieceWriter.Commit(ctx, info)
-				info := &pieces.Info{
-					SatelliteID: limit.SatelliteId,
-
-					PieceID:         limit.PieceId,
-					PieceSize:       pieceWriter.Size(),
-					PieceCreation:   limit.OrderCreation,
-					PieceExpiration: limit.PieceExpiration,
-
-					OrderLimit:      limit,
-					UplinkPieceHash: message.Done,
+				var expTime *time.Time
+				if !limit.PieceExpiration.IsZero() {
+					expTime = &limit.PieceExpiration
 				}
-
-				if err := endpoint.pieceinfo.Add(ctx, info); err != nil {
-					ignoreCancelContext := context.Background()
-					deleteErr := endpoint.store.Delete(ignoreCancelContext, limit.SatelliteId, limit.PieceId)
-					return ErrInternal.Wrap(errs.Combine(err, deleteErr))
+				info := &pb.PieceHeader{
+					Hash:           calculatedHash,
+					CreationTime:   message.Done.Timestamp,
+					ExpirationTime: expTime,
+					Signature:      message.Done.GetSignature(),
+					OrderLimit:     *limit,
+				}
+				if err := pieceWriter.Commit(ctx, info); err != nil {
+					return ErrInternal.Wrap(err) // TODO: report grpc status internal server error
```

Comment:
These "TODO: report grpc status" comments are sprinkled very liberally throughout our codebase, and I don't even know what they mean. I don't think it's a trivial transformation, or presumably it would have been done already. I just kept this comment intact from the previous code.

Reply:
It's trivial

```golang
	return status.Error(code.Internal, "blablabal")
```

but if you keep it from the previous code, then it's fair to leave it as it's, I thought that it was a new code.

Reply:
Oh! I'll do that then. Thanks

Reply:
Ok, but remember to log the error before returning it because it's an internal error and it's important to get as much info as possible and when we return the gRPC error we lose the info of `err` (stack trace, etc.).

Reply:
I'm actually going to elect to keep these TODOs in place for now, since this PR needs to be pushed through. But thanks for educating me about that.